Use 'txnedit' task instead of 'simpleEdit' for updating transactions

ScottG489 opened the pull request:
The 'txnedit' task is able to update all properties of a transaction
including notes, tags, etc. It is also compatible with all fields
which can be submitted with the 'simpleEdit' task.


Additionally I'd like to point out that with your args schema this should all be compatible. All of the field names are the same for both types of tasks. The 'simpleEdit' task seems to be used when you edit a transaction from the "Transactions" page on mint directly in the list of transactions. If you select a transaction in the transaction list and click on the "Edit details" tab just below the row, it expands the transaction and allows you to edit more details of the transaction. Clicking "Update" then posts the transaction update using the 'txnedit' task.

Let me know if you have any questions and I'd be happy to discuss this PR in more detail. I plan on updating the schema further to have explicit support for things like notes and tags if that's alright with you. However, for now I'd just like to make this small, nonbreaking change that will allow for future additions to the schema. Thanks!

dhleong commented:
Hey, thanks for the PR!

I vaguely remember running into some issue trying using `txnedit` for my
use case. Have you tested it to make sure omitting fields doesn't set them
to blank, for example?

On Tue, May 8, 2018 at 2:48 AM Scott G <notifications@github.com> wrote:

> The 'txnedit' task is able to update all properties of a transaction
> including notes, tags, etc. It is also compatible with all fields
> which can be submitted with the 'simpleEdit' task.
>
> Additionally I'd like to point out that with your args schema this should
> all be compatible. All of the field names are the same for both types of
> tasks. The 'simpleEdit' task seems to be used when you edit a transaction
> from the "Transactions" page on mint directly in the list of transactions.
> If you select a transaction in the transaction list and click on the "Edit
> details" tab just below the row, it expands the transaction and allows you
> to edit more details of the transaction. Clicking "Update" then posts the
> transaction update using the 'txnedit' task.
>
> Let me know if you have any questions and I'd be happy to discuss this PR
> in more detail. I also plan on updating the schema further to have explicit
> support for things like notes and tags if that's alright with you. However,
> for now these attributes can be added dynamically to the args parameter
> which will suit my immediate needs. Thanks!
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/dhleong/pepper-mint/pull/14
> Commit Summary
>
>    - Use 'txnedit' task instead of 'simpleEdit' for updating transactions
>
> File Changes
>
>    - *M* index.js
>    <https://github.com/dhleong/pepper-mint/pull/14/files#diff-0> (2)
>
> Patch Links:
>
>    - https://github.com/dhleong/pepper-mint/pull/14.patch
>    - https://github.com/dhleong/pepper-mint/pull/14.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dhleong/pepper-mint/pull/14>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAx0FiJNkb49dSkRCX1wmbOBHO_NDqrmks5twT_AgaJpZM4T2GmK>
> .
>

ScottG489 commented:
I'm adding some tests to this PR which cover all of the supported fields. The currently supported fields exposed via args appear to be:

category
catId
date
merchant
txnId

My test proves all of those fields are still updated with the new task. Perhaps they changed their API since you last tested out the 'txnedit' task?

dhleong commented on test/integration-test.js:
```diff
@@ -0,0 +1,72 @@
+  var config = require('./integration-config.json');
+} catch (e) {
+  // no config; don't run tests
+  process.exit()
```
When running `npm test`, won't this prevent it from running other tests?

ScottG489 replied:
You're right! I was hoping for a more elegant solution than what was in the other test with it all wrapped in a big conditional block but I suppose I'll just do it that way for now.

dhleong commented on test/integration-test.js:
```diff
@@ -0,0 +1,73 @@
+  let editTransactionRequest = {
+    id: tnx[0].id,
+    merchant: "New Test Merchant Name",
+    category: "Vacation",
```
It's been a while so I don't remember what all fields are accepted by this API, but for example if `notes` was set for the transaction, would this request delete that now that it's using `txnedit`?

ScottG489 replied:
You're right, other fields such as notes, tags, etc. are allowed with this task. However, I was just hoping to validate changing this task would be fine on this PR and add a new schema for the args in another PR if that's alright with you. Just wanted to keep it small. Since your schema on the input args is restricted (people can't just add whatever fields they want) it should still work fine. Does that make sense?

ScottG489 replied:
I see what you're saying. In your previous experience, any fields which you did not send were erased. I was actually worried about this as well but it does not seem to be the case. Let me see if I can write up a test case to prove this.

ScottG489 replied:
I've implemented these tests.

ScottG489 commented:
Let me know if you have any other questions about my PR. I pushed a change addressing your concern about the `process.exit()` issue and responded to your other comments. Thanks!

ScottG489 commented:
I just pushed a change that mostly addresses your concern about missing fields. I have a few things I'd like to expand on, however.

Firstly, the test which sends in undefined fields has been ignored. This is due to what appears to be a bug in PepperMint.editTransaction() which requires you to send in a date.

Second, you might notice that this doesn't prove that fields such as note, tag, etc won't be overwritten if not specified. As mentioned before, since these fields cannot yet be set with the current schema of args passed to editTransaction() then we're are safe. Once I update the schema of args (in a following PR) to allow for things such as note, tag, etc. then I will add those fields to the tests to prove they are unaffected if not specified as well.

Hopefully this addresses your concerns. Let me know what you think. Thanks!

ScottG489 commented on test/integration-test.js:
```diff
@@ -0,0 +1,145 @@
+let chai = require('chai');
+let PepperMint = require('../index')
```
Fixed!

ScottG489 commented:
Ok, I think I understand what you're saying. It doesn't matter that the user can't pass in something like a note. What matters is that when pepper-mint talks to the backing mint API it's not passing certain fields and this could possibly be construed by the API as them being "empty" and then it clearing those fields.

I've added an assertion that the note remains unchanged. Let me know if this is what you were looking for! I can add in assertions on other fields on top of just note to verify they haven't changed as well if you'd like.

dhleong commented:
Okay great, thanks! Last are just a few style nits. I apologize that the `.eslintrc` files aren't up to date, but I don't see too many contributors to this project and rarely work on it myself these days.

* I prefer semicolons at the end of lines
* There should be no spaces between `function` and the opening paren of the parameter list
* Please include a trailing comma in maps and lists, eg: [here](https://github.com/dhleong/pepper-mint/pull/14/files#diff-1d067989ae31aa7e3c57eeaa93b61874R85). (The main project uses leading commas, but I'd like to migrate off of that at some point to trailing commas. Alas, all my old javascript projects use those leading commas....)
* I prefer 4-space indents, but if you don't have an easy way to reformat that then don't worry about it for the test file.

ScottG489 commented:
Pushed a few changes to address your comments.

dhleong commented:
You rock, man 👍

ScottG489 commented:
Thanks!
